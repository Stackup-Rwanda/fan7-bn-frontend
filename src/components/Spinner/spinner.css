.loading {
  width: 100px;
  height: 100px;
  margin: 22% auto;
  -webkit-box-sizing: border-box;
          box-sizing: border-box;
  border-radius: 50%;
  border-top: 5px solid #A2AAB1;
  position: relative;
  vertical-align: middle;
  -webkit-animation: a1 2s linear infinite;
          animation: a1 2s linear infinite;
}

.loading::before, .loading::after {
  content: '';
  width: 100px;
  height: 100px;
  position: absolute;
  left: 0;
  top: -5px;
  -webkit-box-sizing: border-box;
          box-sizing: border-box;
  border-top: 5px solid #A2AAB1;
  border-radius: 50%;
}

.loading::before {
  -webkit-transform: rotate(120deg);
          transform: rotate(120deg);
}

.loading::after {
  -webkit-transform: rotate(240deg);
          transform: rotate(240deg);
}

.loading span {
  position: absolute;
  width: 100px;
  height: 100px;
  color: #000000;
  -webkit-box-sizing: border-box;
          box-sizing: border-box;
  text-align: center;
  line-height: 100px;
  -webkit-animation: a2 2s linear infinite;
          animation: a2 2s linear infinite;
}

@-webkit-keyframes a1 {
  to {
    -webkit-transform: rotate(360deg);
            transform: rotate(360deg);
  }
}

@keyframes a1 {
  to {
    -webkit-transform: rotate(360deg);
            transform: rotate(360deg);
  }
}

@-webkit-keyframes a2 {
  to {
    -webkit-transform: rotate(-360deg);
            transform: rotate(-360deg);
  }
}

@keyframes a2 {
  to {
    -webkit-transform: rotate(-360deg);
            transform: rotate(-360deg);
  }
}
/*# sourceMappingURL=spinner.css.map */