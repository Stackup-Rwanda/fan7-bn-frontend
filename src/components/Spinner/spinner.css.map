{
    "version": 3,
    "mappings": "AAAA,AAAA,QAAQ,CAAC;EACL,KAAK,EAAE,KAAK;EACZ,MAAM,EAAE,KAAK;EACb,MAAM,EAAE,QAAQ;EAChB,UAAU,EAAE,UAAU;EACtB,aAAa,EAAE,GAAG;EAClB,UAAU,EAAE,iBAAiB;EAC7B,QAAQ,EAAE,QAAQ;EAClB,cAAc,EAAE,MAAM;EACtB,SAAS,EAAE,qBAAqB;CAiCnC;;AA1CD,AAWI,QAXI,AAWH,QAAQ,EAXb,QAAQ,AAYH,OAAO,CAAC;EACL,OAAO,EAAE,EAAE;EACX,KAAK,EAAE,KAAK;EACZ,MAAM,EAAE,KAAK;EACb,QAAQ,EAAE,QAAQ;EAClB,IAAI,EAAE,CAAC;EACP,GAAG,EAAE,IAAI;EACT,UAAU,EAAE,UAAU;EACtB,UAAU,EAAE,iBAAiB;EAC7B,aAAa,EAAE,GAAG;CACrB;;AAtBL,AAwBI,QAxBI,AAwBH,QAAQ,CAAC;EACN,SAAS,EAAE,cAAc;CAC5B;;AA1BL,AA4BI,QA5BI,AA4BH,OAAO,CAAC;EACL,SAAS,EAAE,cAAc;CAC5B;;AA9BL,AAgCI,QAhCI,CAgCJ,IAAI,CAAC;EACD,QAAQ,EAAE,QAAQ;EAClB,KAAK,EAAE,KAAK;EACZ,MAAM,EAAE,KAAK;EACb,KAAK,EAAE,OAAO;EACd,UAAU,EAAE,UAAU;EACtB,UAAU,EAAE,MAAM;EAClB,WAAW,EAAE,KAAK;EAClB,SAAS,EAAE,qBAAqB;CACnC;;AAGL,UAAU,CAAV,EAAU;EACN,EAAE;IACE,SAAS,EAAE,cAAc;;;;AAIjC,UAAU,CAAV,EAAU;EACN,EAAE;IACE,SAAS,EAAE,eAAe",
    "sources": [
        "spinner.scss"
    ],
    "names": [],
    "file": "spinner.css"
}