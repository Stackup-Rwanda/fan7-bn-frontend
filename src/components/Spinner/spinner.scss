.loading {
    width: 100px;
    height: 100px;
    margin: 22% auto;
    box-sizing: border-box;
    border-radius: 50%;
    border-top: 5px solid #A2AAB1;
    position: relative;
    vertical-align: middle;
    animation: a1 2s linear infinite;

    &::before, 
    &::after {
        content: '';
        width: 100px;
        height: 100px;
        position: absolute;
        left: 0;
        top: -5px;
        box-sizing: border-box;
        border-top: 5px solid #A2AAB1;
        border-radius: 50%;
    }

    &::before {
        transform: rotate(120deg);
    }

    &::after {
        transform: rotate(240deg);
    }

    span {
        position: absolute;
        width: 100px;
        height: 100px;
        color: #000000;
        box-sizing: border-box;
        text-align: center;
        line-height: 100px;
        animation: a2 2s linear infinite;
    }
}

@keyframes a1 {
    to {
        transform: rotate(360deg);
    }
}

@keyframes a2 {
    to {
        transform: rotate(-360deg);
    }
}