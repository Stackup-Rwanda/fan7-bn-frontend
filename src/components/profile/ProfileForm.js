import React, { Component } from 'react';
import PropTypes from 'prop-types';
import moment from 'moment';
import './styles/profileForm.scss';

class ProfileForm extends Component {
  constructor(props) {
    super(props);
    this.initialState = {
      user: null,
    };
    this.state = this.initialState;
    this.handleSubmit = this.handleSubmit.bind(this);
    this.handleChange = this.handleChange.bind(this);
  }

  handleSubmit(event) {
    event.preventDefault();
    const { updateProfile } = this.props;
    const { user } = this.state;
    updateProfile(user);
  }

  handleChange(event) {
    const { name, value } = event.target;
    this.setState(prevState => ({
      user: {
        ...prevState.user,
        [name]: value,
      },
    }));
  }

  render() {
    const { profile, loading } = this.props;
    const {
      first_name: firstName,
      last_name: lastName,
      user_name: userName,
      gender,
      dob,
      phone,
      address,
      country,
      prefered_language: preferedLanguage,
      prefered_currency: preferdCurrency,
      company,
      department,
    } = profile;

    return (
      <form className="form" onSubmit={this.handleSubmit}>
        <div className="form_header">
          <h1 className="form_header__title">Basic info</h1>
          <div className="form_header__buttons">
            <button type="button" className="form_header__buttons___button button_white">
              Cancel
            </button>
            <button type="submit" className="form_header__buttons___button button_blue">
              {loading ? 'Updating...' : 'Update'}
            </button>
          </div>
        </div>
        <div className="form_content">
          <div className="form_content__inputField">
            <label htmlFor="first_name">
              <span className="form_content__inputField___label">FIRST NAME</span>
              <input
                type="text"
                className="form_content__inputField___input"
                name="first_name"
                defaultValue={firstName}
                placeholder="Enter your first name"
                onChange={this.handleChange}
              />
            </label>
          </div>
          <div className="form_content__inputField">
            <label htmlFor="last_name">
              <span className="form_content__inputField___label">LAST NAME</span>
              <input
                type="text"
                className="form_content__inputField___input"
                name="last_name"
                defaultValue={lastName}
                placeholder="Enter your last name"
                onChange={this.handleChange}
              />
            </label>
          </div>
          <div className="form_content__inputField">
            <label htmlFor="user_name">
              <span className="form_content__inputField___label">USERNAME</span>
              <input
                type="text"
                className="form_content__inputField___input"
                name="user_name"
                defaultValue={userName}
                placeholder="Enter your user name"
                onChange={this.handleChange}
              />
            </label>
          </div>
          <div className="form_content__inputField">
            <label htmlFor="gender">
              <span className="form_content__inputField___label">GENDER</span>
              <div className="form_content__inputField__select">
                <select
                  className="form_content__inputField___input"
                  name="gender"
                  defaultValue={gender}
                  onChange={this.handleChange}
                >
                  <option disabled selected className="placeholder">
                    Select your gender
                  </option>
                  <option value="Male">Male</option>
                  <option value="Female">Female</option>
                </select>
              </div>
            </label>
          </div>
          <div className="form_content__inputField">
            <label htmlFor="dob">
              <span className="form_content__inputField___label">DATE OF BIRTH</span>
              <input
                type="date"
                className="form_content__inputField___input"
                name="dob"
                defaultValue={moment(dob).format('YYYY-MM-DD')}
                onChange={this.handleChange}
              />
            </label>
          </div>
          <div className="form_content__inputField">
            <label htmlFor="phone">
              <span className="form_content__inputField___label">PHONE</span>
              <input
                type="text"
                className="form_content__inputField___input"
                name="phone"
                defaultValue={phone}
                placeholder="Enter your Phone number"
                onChange={this.handleChange}
              />
            </label>
          </div>
          <div className="form_content__inputField">
            <label htmlFor="address">
              <span className="form_content__inputField___label">ADDRESS</span>
              <input
                type="text"
                className="form_content__inputField___input"
                name="address"
                defaultValue={address}
                placeholder="Enter your current address"
                onChange={this.handleChange}
              />
            </label>
          </div>
          <div className="form_content__inputField">
            <label htmlFor="country">
              <span className="form_content__inputField___label">COUNTRY</span>
              <input
                type="text"
                className="form_content__inputField___input"
                name="country"
                defaultValue={country}
                placeholder="Enter your country"
                onChange={this.handleChange}
              />
            </label>
          </div>
          <div className="form_content__inputField">
            <label htmlFor="prefered_language">
              <span className="form_content__inputField___label">PREFERED LANGUAGE</span>
              <input
                type="text"
                className="form_content__inputField___input"
                name="prefered_language"
                defaultValue={preferedLanguage}
                placeholder="Enter your prefered language"
                onChange={this.handleChange}
              />
            </label>
          </div>
          <div className="form_content__inputField">
            <label htmlFor="prefered_currency">
              <span className="form_content__inputField___label">PREFERED CURRENCY</span>
              <input
                type="text"
                className="form_content__inputField___input"
                name="prefered_currency"
                defaultValue={preferdCurrency}
                placeholder="Enter your prefered currency"
                onChange={this.handleChange}
              />
            </label>
          </div>
          <div className="form_content__inputField">
            <label htmlFor="company">
              <span className="form_content__inputField___label">COMPANY</span>
              <input
                type="text"
                className="form_content__inputField___input"
                name="company"
                defaultValue={company}
                placeholder="Enter your company"
                onChange={this.handleChange}
              />
            </label>
          </div>
          <div className="form_content__inputField">
            <label htmlFor="department">
              <span className="form_content__inputField___label">DEPARTMENT</span>
              <input
                type="text"
                className="form_content__inputField___input"
                name="department"
                defaultValue={department}
                placeholder="Enter your department"
                onChange={this.handleChange}
              />
            </label>
          </div>
        </div>
      </form>
    );
  }
}

ProfileForm.propTypes = {
  updateProfile: PropTypes.func.isRequired,
  loading: PropTypes.bool,
  profile: PropTypes.objectOf(PropTypes.any),
};

ProfileForm.defaultProps = {
  loading: false,
  profile: {},
};

export default ProfileForm;
