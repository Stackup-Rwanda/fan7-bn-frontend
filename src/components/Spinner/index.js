import React from 'react';
import './spinner.scss';

const Loading = () => (
  <div className="loading">
    <span>Loading...</span>
  </div>
);

export default Loading;
