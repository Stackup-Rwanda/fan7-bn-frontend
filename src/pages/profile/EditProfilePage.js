import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import Header from '../../components/profile/EditProfileHeader';
import SideBar from '../../components/profile/EditProfileSideBar';
import ProfileForm from '../../components/profile/ProfileForm';
import DashboardLayout from '../../components/DashboardLayout';
import Spinner from '../../components/Spinner';
import profileActions from '../../store/modules/profile/actions';
import selectors from '../../store/modules/profile/selectors';
import './styles/editProfile.scss';
import Chatbot from '../../components/Chatbot';

class EditProfilePage extends Component {
  constructor(props) {
    super(props);
    this.initialState = {
      image: null,
    };
    this.state = this.initialState;
    this.handleUpdateProfile = this.handleUpdateProfile.bind(this);
    this.handleImg = this.handleImg.bind(this);
  }

  componentDidMount() {
    const { dispatch } = this.props;
    const { getProfile } = profileActions;
    dispatch(getProfile());
  }

  handleUpdateProfile(data) {
    const { image } = this.state;

    if (data || image) {
      const { dispatch } = this.props;
      const { updateProfile } = profileActions;
      const dataToUpdate = !image ? data : { ...data, image };

      const formData = new FormData();

      Object.entries(dataToUpdate).forEach(([key, value]) => {
        formData.append(key, value);
      });

      dispatch(updateProfile(formData));
    }
  }

  handleImg(image) {
    this.setState({
      image,
    });
  }

  render() {
    const { loading, updateLoading, profile, email, role, profileImg } = this.props;

    return loading ? (
      <Spinner />
    ) : (
      <div className="editProfile">
        {profile && (
          <>
            <Header />
            <SideBar
              email={email}
              role={role}
              profileImg={profileImg}
              handleChange={this.handleImg}
            />
            <ProfileForm
              profile={profile}
              loading={updateLoading}
              updateProfile={this.handleUpdateProfile}
            />
          </>
        )}
        <Chatbot />
      </div>
    );
  }
}

EditProfilePage.propTypes = {
  dispatch: PropTypes.func.isRequired,
  loading: PropTypes.bool.isRequired,
  updateLoading: PropTypes.bool.isRequired,
  profile: PropTypes.objectOf(PropTypes.any).isRequired,
  email: PropTypes.string.isRequired,
  role: PropTypes.string.isRequired,
  profileImg: PropTypes.string.isRequired,
};

const {
  selectLoading,
  selectLoadingUpdateProfile,
  selectProfile,
  selectProfileEmail,
  selectProfileRole,
  selectProfileAvatar,
} = selectors;

const select = ({ userProfile }) => ({
  loading: selectLoading(userProfile),
  updateLoading: selectLoadingUpdateProfile(userProfile),
  profile: selectProfile(userProfile),
  email: selectProfileEmail(userProfile),
  role: selectProfileRole(userProfile),
  profileImg: selectProfileAvatar(userProfile),
});

export default connect(select)(DashboardLayout(EditProfilePage));
