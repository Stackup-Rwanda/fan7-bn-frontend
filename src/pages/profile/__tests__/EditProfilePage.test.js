/* eslint-disable react/jsx-props-no-spreading */
import React from 'react';
import { mount, shallow } from 'enzyme';
import thunk from 'redux-thunk';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import { BrowserRouter as Router } from 'react-router-dom';
import EditProfilePage from '../EditProfilePage';
import Spinner from '../../../components/Spinner';
import Sidebar from '../../../components/profile/EditProfileSideBar';
import ProfileForm from '../../../components/profile/ProfileForm';

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

const history = {
  push: jest.fn(),
  replace: jest.fn(),
  createHref: jest.fn(),
};

describe('EditProfilePage Component Test', () => {
  it('should return true if component exists', () => {
    localStorage.setItem('barefoot_nomad_token', 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpZCI6NiwiZW1haWwiOiJlbHZpc3J1Z2FtYmFAZ21haWwuY29tIiwicm9sZSI6Im1hbmFnZXIiLCJpbWFnZV91cmwiOm51bGwsImlhdCI6MTU4NjYyODg1OH0.i8T-qF8n0e4jySQkaP1G7bUG9oMmiL1BC6DYiJHknLE');

    const initialState = {
      chat: {
        messages: {
          senderId: 1,
          sender: 'The Great',
          message: 'Test react',
          createdAt: '01-03-2019'
        }
    }};
    const store = mockStore(initialState);
    const wrapper = mount(
      <Provider store={store}>
        <Router history={history}>
          <EditProfilePage />
        </Router>
      </Provider>,
    );

    expect(wrapper.find('.editProfile').exists()).toBe(true);
  });

  it("should display spinner if it's loading", () => {
    localStorage.setItem('barefoot_nomad_token', 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpZCI6NiwiZW1haWwiOiJlbHZpc3J1Z2FtYmFAZ21haWwuY29tIiwicm9sZSI6Im1hbmFnZXIiLCJpbWFnZV91cmwiOm51bGwsImlhdCI6MTU4NjYyODg1OH0.i8T-qF8n0e4jySQkaP1G7bUG9oMmiL1BC6DYiJHknLE');

    const initialState = {
      userProfile: { loading: true },
      chat: {
        messages: {
          senderId: 1,
          sender: 'The Great',
          message: 'Test react',
          createdAt: '01-03-2019'
        }
      }
    };
    const store = mockStore(initialState);
    const wrapper = mount(
      <Provider store={store}>
        <Router history={history}>
          <EditProfilePage />
        </Router>
      </Provider>,
    );

    expect(wrapper.find(Spinner).exists()).toBe(true);
  });

  it('should display Sidebar if it is not loading', () => {
    localStorage.setItem('barefoot_nomad_token', 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpZCI6NiwiZW1haWwiOiJlbHZpc3J1Z2FtYmFAZ21haWwuY29tIiwicm9sZSI6Im1hbmFnZXIiLCJpbWFnZV91cmwiOm51bGwsImlhdCI6MTU4NjYyODg1OH0.i8T-qF8n0e4jySQkaP1G7bUG9oMmiL1BC6DYiJHknLE');

    const handleUpdateProfile = jest.fn();
    const data = { user_name: 'elvisR' };
    const image = 'img.jpg';
    const initialState = {
      userProfile: { loading: false, profile: {} },
      chat: {
        messages: {
          senderId: 1,
          sender: 'The Great',
          message: 'Test react',
          createdAt: '01-03-2019'
        }
      }
    };
    const store = mockStore(initialState);
    const wrapper = mount(
      <Provider store={store}>
        <Router history={history}>
          <EditProfilePage />
        </Router>
      </Provider>,
    );

    expect(wrapper.find(ProfileForm).exists()).toBe(true);

    wrapper
      .find(ProfileForm)
      .props()
      .updateProfile(data);
    wrapper
      .find(ProfileForm)
      .props()
      .updateProfile();
    wrapper
      .find(Sidebar)
      .props()
      .handleChange(image);
  });
});

describe('<EditProfilePage /> shallow rendering tests', () => {
  const initialState = {};
  const store = mockStore(initialState);

  it('matches the snapshot', () => {
    const tree = shallow(<EditProfilePage store={store} />);
    expect(tree).toMatchSnapshot();
  });
});
