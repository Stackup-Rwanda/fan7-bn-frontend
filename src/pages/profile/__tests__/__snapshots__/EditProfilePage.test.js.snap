// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EditProfilePage /> shallow rendering tests matches the snapshot 1`] = `
<ContextProvider
  value={null}
>
  <_class
    dispatch={[Function]}
    email=""
    loading={false}
    profile={null}
    profileImg=""
    role=""
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
    updateLoading={false}
  />
</ContextProvider>
`;
